### PR TITLE
Remove calculation of unnecessary matrix columns in SVD gradient

### PR DESCRIPTION
**Pull request by vnavkal**
The SVD gradient calculation when `compute_uv=False` currently uses the orthogonal "U" and "V" matrices returned by the SVD operation with `full_matrices=True`, but it really requires only the `full_matrices=False` versions.  This pull request makes the calculation use the `full_matrices=False` versions.

@rmlarsen pointed out that this change could be made in https://github.com/tensorflow/tensorflow/pull/14259#discussion_r157067512.

### CODE REVIEW

**Comment by vnavkal** [comment]
If desired, I could also make some other simplifications to the code that calculates the gradient when `compute_uv=False`:
1. Remove the conditional on `use_adjoint`: the formula is really the same whether `use_adjoint` is `True` or `False`.
2. Move it outside the `with ops.control_dependencies([grad_s, grad_u, grad_v])` block, perhaps to a new `with ops.control_dependencies([grad_s])` block.  (It doesn't use `grad_u` or `grad_v`.)  It's not clear to me why `ops.control_dependencies` is needed here at all, though, so I won't do this unless someone else confirms that it makes sense to do it.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rmlarsen** [comment]
@vnavkal Thanks for the contribution and sorry for the delay. It would be wonderful if you could implement the proposed improvements to the gradients in this PR as well. What do you think?

File: tensorflow/python/ops/linalg_grad.py [hunk 1]
```diff
@@ -301,7 +301,7 @@ def _SvdGrad(op, grad_s, grad_u, grad_v):
         "inner matrix shape.")
 
   if not op.get_attr("compute_uv"):
-    s, u, v = linalg_ops.svd(a, compute_uv=True, full_matrices=True)
+    s, u, v = linalg_ops.svd(a, compute_uv=True)
```
Comment (rmlarsen) [inline tensorflow/python/ops/linalg_grad.py]: Should we not compute the full matrices if op.get_attr("full_matrices") == True? See line 375..
BTW, I would add:

full_matrices = op.get_attr("full_matrices") 
comput_uv = op.get_attr("compute_uv")

at the top of the function to improve readability.

**Review by rmlarsen — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/linalg_grad.py (hunk 1 above)
Comment (vnavkal) [reply tensorflow/python/ops/linalg_grad.py]: > Should we not compute the full matrices if op.get_attr("full_matrices") == True? See line 375.

There's no need -- if `full_matrices` is true, then the function returns early and never reaches line 375: https://github.com/vnavkal/tensorflow/blob/66a8c03a8eaf99de5eca41e1f7a7bf9d0209539d/tensorflow/python/ops/linalg_grad.py#L321-L329.  The change I suggested in my comment above should make this less confusing; I'll implement it.

> BTW, I would add:
> ```
> full_matrices = op.get_attr("full_matrices") 
> compute_uv = op.get_attr("compute_uv")
> ```
> at the top of the function to improve readability.

sounds good, I'll do this.

**Comment by vnavkal** [comment]
Sure, I'll make the extra changes I suggested.

**Comment by rmlarsen** [comment]
@vnavkal Thanks you, let me know when this is ready for another look.

**Comment by vnavkal** [comment]
I added two commits that implemented the suggestions:
1. 15519c79c7b37533dd447c0abb91985b2c31a187 simplifies and consolidates the logic for the case when `compute_uv` is false.
2. fc109072c64d5ef7b48c1c604d7dca3360984448 assigns `op.get_attr("full_matrices")` to the local variable `full_matrices`.  I didn't create a variable for `op.get_attr("compute_uv")` because it's now used only once.

File: tensorflow/python/ops/linalg_grad.py [hunk 2]
```diff
@@ -277,20 +277,28 @@ def _SvdGrad(op, grad_s, grad_u, grad_v):
...
-    m = a_shape[-2].merge_with(grad_u_shape[-2])
-    n = a_shape[-1].merge_with(grad_v_shape[-2])
-    batch_shape = a_shape[:-2].merge_with(grad_u_shape[:-2]).merge_with(
-        grad_v_shape[:-2])
-    a_shape = batch_shape.concatenate([m, n])
+  if not op.get_attr("compute_uv"):
+    s, u, v = linalg_ops.svd(a, compute_uv=True)
+    grad_a = math_ops.matmul(u, math_ops.matmul(grad_s_mat, v, adjoint_b=True))
+    grad_a.set_shape(a_shape)
+    return grad_a
```
Comment (vnavkal) [inline tensorflow/python/ops/linalg_grad.py]: Note that all new code above here was previously within a `with ops.control_dependencies([grad_s, grad_u, grad_v])` block.  I don't see why those dependencies needed to be declared explicitly, but maybe I'm missing something?

**Comment by vnavkal** [comment]
@rmlarsen this should be ready for another look whenever you get a chance.  Thanks!

File: tensorflow/python/ops/linalg_grad.py (hunk 2 above)
Comment (rmlarsen) [reply tensorflow/python/ops/linalg_grad.py]: I agree that these are probably not needed.

**Review by rmlarsen — APPROVED** [review]

**Pull request merged**